In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns
from scipy import stats

sns.set(style='dark', color_codes=True, font_scale=1.3)

In [2]:
from IPython.core.display import HTML
css = open('style-table.css').read() + open('style-notebook.css').read()
HTML('<style>{}</style>'.format(css))

In [19]:
df_inmates = pd.read_csv('exinmates.csv', encoding="latin-1")
df_inmates = df_inmates.set_index("Execution No.")
df_inmates

TDCJ Number   First Name    Last Name Date Received  \
Execution No.                                                        
545                 999275        Ruben     Cardenas       7/29/98   
544                 999411       Robert       Pruett       4/30/02   
543                 999494      Taichin       Preyor       3/30/05   
542                    997        James        Bigby           NaN   
541                 999145      Rolando         Ruiz           NaN   
540                 999463        Terry      Edwards       12/4/03   
539                 999533  Christopher      Wilkins       3/12/08   
538                 999481       Barney       Fuller       7/21/04   
537                 999297        Pablo      Vasquez       3/30/99   
536                 999525         Adam         Ward           NaN   
535                 999281          Coy     Wesbrook        9/2/98   
534                 999018      Gustavo       Garcia           NaN   
533                 999539        James      Freeman      11/10/08   
532                 999414      Richard    Masterson       5/15/02   
531                 999419      Raphael      Holiday       6/21/02   
530                 999432        Licho    Escamilla       11/1/02   
529                 999360         Juan       Garcia       6/21/00   
528                 999555       Daniel        Lopez       3/16/10   
527                 999430      Gregory      Russeau      10/15/02   
526                    764       Lester        Bower           NaN   
525                 999451      Derrick      Charles       5/27/03   
524                 999434       Manuel        Garza       11/8/02   
523                 999471         Kent      Sprouse        3/1/04   
522                 999336       Manuel      Vasquez      11/30/99   
521                 999403       Donald      Newbury       1/28/02   
520                 999237       Robert         Ladd       8/29/97   
519                 999149       Arnold       Prieto           NaN   
518                 999400       Miguel      Paredes       11/9/01   
517                 999511         Lisa      Coleman       6/22/06   
516                 999085       Willie      Trottie           NaN   
...                    ...          ...          ...           ...   
30                     624         Leon         King           NaN   
29                     738      Raymond  Landry, Sr.           NaN   
28                     546       Donald     Franklin           NaN   
27                     746       Robert    Streetman           NaN   
26                     586       Joseph    Starvaggi           NaN   
25                     610         John     Thompson           NaN   
24                     739       Elliot      Johnson           NaN   
23                     619      Anthony     Williams           NaN   
22                     759       Eliseo       Moreno           NaN   
21                     667        Ramon    Hernandez       9/30/80   
20                     774      Richard      Andrade           NaN   
19                     608      Michael        Evans           NaN   
18                     678      Chester       Wicker           NaN   
17                     643        Larry        Smith           NaN   
16                     646        Randy       Woolls           NaN   
15                     522      Kenneth        Brock           NaN   
14                     606         Rudy     Esquivel           NaN   
13                     686          Jay    Pinkerton           NaN   
12                     714      Jeffery       Barney           NaN   
11                     662      Charles         Bass           NaN   
10                     555      Charles     Rumbaugh           NaN   
9                      551        Henry       Porter           NaN   
8                      628      Charles       Milton           NaN   
7                      713        Jesse   De La Rosa           NaN   
6                      71

In [57]:
pov_rates = pd.read_csv('pov_rates.csv')
county_codes = pd.read_csv('county_codes.csv')

In [58]:
county_codes.head()

County  FIPS #
0  Anderson       1
1   Andrews       3
2  Angelina       5
3   Aransas       7
4    Archer       9

In [59]:
pov_rates.head()

FIPS #  Year  Rate
0       1  1989  20.5
1       1  1993  20.8
2       1  1995  20.1
3       1  1996   NaN
4       1  1997  20.4

In [66]:
pov_by_county = pd.merge(pov_rates, county_codes, on='FIPS #')
pov_by_county = pov_by_county.drop("FIPS #", axis=1).set_index("County")
pov_by_county.head()

Year  Rate
County              
Anderson  1989  20.5
Anderson  1993  20.8
Anderson  1995  20.1
Anderson  1996   NaN
Anderson  1997  20.4

In [4]:
df_inmates['Age Received'].mean()

28.564935064935064

In [5]:
df_inmates['Age at Offense'].mean()

26.954545454545453

Most inmates on average enter death row approximately 2.5 years after their offense (quick turn around time).

In [6]:
df_inmates['Execution Age'].mean()

39.3394495412844

Most inmates on average stay on death row for approximately 11 years before being executed.

In [7]:
df_inmates['Gender'].value_counts()

Male      150
Female      3
male        1
Name: Gender, dtype: int64

Clearly, males outnumber females.

In [8]:
df_inmates['Race'].value_counts()

White       242
Black       198
Hispanic    103
Other         2
Name: Race, dtype: int64

A majority of inmates being White, with a close 2nd of Black.

In [9]:
df_inmates['Education Level'].max()

12.0

In [10]:
df_inmates['Education Level'].min()

3.0

In [11]:
df_inmates['Education Level'].mean()

9.876623376623376

The lowest being 3rd grade and the highest finishing high school, inmates average an education level of 9th grade.

In [12]:
df_inmates['Sentiment'].value_counts()

 1    200
 0    192
-1    153
Name: Sentiment, dtype: int64

A very close distribution between positive, negative, and neutral sentiments.  However, inmates are more positive at the time of their death.

In [13]:
df_inmates['Poverty Rate'].mean()

17.570458715596356

The average poverty county rate in Texas is 17.57%

In [69]:
counties = df_inmates['County']

# counties['Inmates'] = df_inmates['County'].unique()
counties.head()

Execution No.
545    Hidalgo
544        Bee
543      Bexar
542    Tarrant
541      Bexar
Name: County, dtype: object

In [16]:
counties.groupby('County', counties['Poverty Rate']).mean()

TypeError: 'Series' objects are mutable, thus they cannot be hashed

In [ ]:
sns.regplot(x="Poverty Rate", y="Inmates", data=counties)

In [ ]:
np.corrcoef(x=counties["Poverty Rate"], y=counties["Inmates"])

In [ ]:
counties_3 = counties[counties.Inmates > 2]

In [ ]:
np.corrcoef(x=counties_3["Poverty Rate"], y=counties_3["Inmates"])

In [ ]:
counties_10 = counties[counties.Inmates > 9]
np.corrcoef(x=counties_10["Poverty Rate"], y=counties_10["Inmates"])


In [ ]:
sns.regplot(x="Poverty Rate", y="Inmates", data=counties_10)

In [ ]:
x = counties_10['Poverty Rate']
y = counties_10['Inmates']
slope, intercept, r_value, p_value, std_err = stats.linregress(x,y)

In [ ]:
print (slope, intercept, r_value, p_value, std_err)

In [ ]:
counties.head()

In [ ]:
import bokeh.sampledata
from bokeh.io import show
from bokeh.models import (
    ColumnDataSource,
    HoverTool,
    LogColorMapper
)
from bokeh.palettes import Viridis6 as palette
from bokeh.plotting import figure

from bokeh.sampledata.us_counties import data as tx_counties
# from bokeh.sampledata.unemployment import data as unemployment

palette.reverse()

tx_counties = {
    code: county for code, county in tx_counties.items() if county["state"] == "tx"
}

county_xs = [county["lons"] for county in tx_counties.values()]
county_ys = [county["lats"] for county in tx_counties.values()]

county_names = [county['name'] for county in tx_counties.values()]

for n in county_names:
    try:
        print(counties.ix[n]['Poverty Rate'])
        print("__")
    except:
        print("not found")
        print("---")

county_rates = [counties.ix[name]['Poverty Rate'] if name in counties else 0 for name in county_names]


In [ ]:
color_mapper = LogColorMapper(palette=palette)

source = ColumnDataSource(data=dict(
    x=county_xs,
    y=county_ys,
    name=county_names,
    rate=county_rates,
))

TOOLS = "pan,wheel_zoom,reset,hover,save"

p = figure(
    title="Texas Executed Inmates", tools=TOOLS,
    x_axis_location=None, y_axis_location=None
)
p.grid.grid_line_color = None

p.patches('x', 'y', source=source,
          fill_color={'field': 'rate', 'transform': color_mapper},
          fill_alpha=0.7, line_color="white", line_width=0.5)

hover = p.select_one(HoverTool)
hover.point_policy = "follow_mouse"
hover.tooltips = [
    ("Name", "@name"),
    ("Poverty rate)", "@rate%"),
    ("(Long, Lat)", "($x, $y)"),
]

show(p)